File: banc/clustering/banc-spectral-clustering.py
#!/usr/bin/env python3
"""
Spectral Clustering of BANC CNS Network

Port of python/spectral_clustering/banc_spectral_clustering_final.ipynb
for use on O2 (SLURM) and locally.

Algorithm:
  1. Load BANC meta + simple edgelist from feather (source=v2 or v3)
  2. Quality filter (exclude glia, trachea, unproofread, etc.)
  3. Assign clustering_set (central brain, VNC, neck, visual)
  4. Filter edgelist, prune to strongly connected component
  5. Build column-normalized, symmetrized adjacency matrix
  6. Spectral clustering (normalized Laplacian, bottom-k eigenvectors, KMeans)
  7. UMAP of spectral embedding
  8. Assign cns_network labels (from local meta or GCS via --gcs)
  9. Save CSV (output suffix carries the source: _v2 or _v3)

Parameters (paper Methods §"Spectral clustering"):
  min_connection_strength = 1
  cluster_count = 13
  cluster_seed = 10
  embedding_seed = 3
  UMAP: n_neighbors=100, metric=cosine, min_dist=0, n_components=2

Usage:
  python banc-spectral-clustering.py --data-dir path/to/banc_888/ --source v3
  python banc-spectral-clustering.py  # auto-detects O2 vs local, source=v3
"""

import argparse
import os
import subprocess
import sys
import tempfile
import time

import numpy as np
import pandas as pd
from scipy.optimize import linear_sum_assignment
from scipy.sparse import coo_matrix, csc_matrix
from scipy.sparse.csgraph import laplacian
from scipy.sparse.linalg import eigsh
from sklearn.cluster import KMeans
from sklearn.preprocessing import normalize
import umap
import plotly.express as px


# ---------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------

def parse_args():
    parser = argparse.ArgumentParser(
        description="Spectral clustering of BANC CNS connectivity network"
    )
    parser.add_argument(
        "--data-dir", type=str, default=None,
        help="Directory containing banc_{version}_meta.feather and "
             "banc_{version}_edgelist_simple_{v2|v3}.feather. "
             "Auto-detects O2 vs local if not specified."
    )
    parser.add_argument(
        "--source", type=str, default=None, choices=[None, "v2", "v3"],
        help="Synapse source for the edgelist suffix. "
             "Defaults to $BANC_SYN_SOURCE or 'v3'."
    )
    parser.add_argument(
        "--output-dir", type=str,
        default=os.path.join(
            os.environ.get(
                "BANC_PROJECT_DATA_DIR",
                "/n/data1/hms/neurobio/wilson/banc/BANC-project/data"),
            "cns_network"),
        help=("Output directory for clustering CSV. Default: "
              "$BANC_PROJECT_DATA_DIR/cns_network, falling back to "
              "/n/data1/.../BANC-project/data/cns_network. Set "
              "BANC_PROJECT_DATA_DIR to redirect to a different repo checkout "
              "(e.g. on another machine).")
    )
    parser.add_argument(
        "--gcs", action="store_true", default=False,
        help="Fetch cns_network labels from GCS meta feather "
             "(gs://lee-lab_brain-and-nerve-cord-fly-connectome/compiled_data/banc_<ver>/) "
             "via gsutil. If not set, uses cns_network from the local meta."
    )
    parser.add_argument(
        "--min-connection-strength", type=int, default=1,
        help="Minimum synapse count per connection (default: 1, paper Methods)"
    )
    parser.add_argument(
        "--cluster-count", type=int, default=13,
        help="Number of spectral clusters (default: 13, paper Methods)"
    )
    parser.add_argument(
        "--cluster-seed", type=int, default=10,
        help="Random seed for KMeans (default: 10)"
    )
    parser.add_argument(
        "--embedding-seed", type=int, default=3,
        help="Random seed for UMAP (default: 3)"
    )
    parser.add_argument(
        "--banc-version", type=int, default=888,
        help="BANC dataset version (default: 888)"
    )
    args = parser.parse_args()
    if args.source is None:
        args.source = os.environ.get("BANC_SYN_SOURCE", "v3").lower()
    if args.source not in ("v2", "v3"):
        parser.error(f"--source must be 'v2' or 'v3' (got {args.source!r})")
    return args


def detect_data_dir(version):
    """Auto-detect data directory: O2 cluster vs local."""
    candidates = [
        # O2 HPC
        f"/n/data1/hms/neurobio/wilson/connectomes/banc/banc_{version}",
        # Local (relative to repo root)
        f"lee-lab_brain-and-nerve-cord-fly-connectome/compiled_data/banc_{version}",
    ]
    for path in candidates:
        if os.path.isdir(path):
            return path
    print(f"ERROR: Could not find data directory. Tried: {candidates}",
          file=sys.stderr)
    sys.exit(1)


# ---------------------------------------------------------------
# Spectral clustering
# ---------------------------------------------------------------

def spectral_clustering_from_adj(adj_sym, num_clusters, random_state=None):
    """
    Spectral clustering on a symmetric adjacency matrix.

    1. Compute normalized Laplacian
    2. Find bottom-k eigenvectors (smallest eigenvalues)
    3. Row-normalize the eigenvector embedding
    4. KMeans clustering

    Returns (labels, embedding) where labels are 1-indexed.
    """
    # Normalized Laplacian: L = I - D^{-1/2} A D^{-1/2}
    lap = laplacian(adj_sym, normed=True)

    # Bottom-k eigenvectors
    eigval, eigvec = eigsh(lap, k=num_clusters, which="SM")

    # Row-normalize
    embedding = normalize(eigvec)

    # KMeans
    km = KMeans(n_clusters=num_clusters, random_state=random_state, n_init=10)
    labels = km.fit_predict(embedding) + 1  # 1-indexed

    return labels, embedding


def perform_umap(embedding, n_neighbors=100, random_state=3):
    """UMAP projection of spectral embedding to 2D."""
    model = umap.UMAP(
        n_neighbors=n_neighbors,
        metric="cosine",
        min_dist=0,
        n_components=2,
        random_state=random_state,
        n_jobs=1,
    )
    return model.fit_transform(embedding)


# ---------------------------------------------------------------
# Main pipeline
# ---------------------------------------------------------------

def main():
    args = parse_args()
    t_start = time.time()

    # Resolve data directory
    data_dir = args.data_dir or detect_data_dir(args.banc_version)
    print(f"Data directory: {data_dir}")

    # --- 1. Load data ---
    print("Loading meta and edgelist...")
    meta_path = os.path.join(data_dir, f"banc_{args.banc_version}_meta.feather")
    el_path = os.path.join(
        data_dir,
        f"banc_{args.banc_version}_edgelist_simple_{args.source}.feather",
    )
    meta = pd.read_feather(meta_path)
    edgelist = pd.read_feather(el_path)
    print(f"  Meta: {meta.shape[0]} neurons, Edgelist: {edgelist.shape[0]} edges")

    # Use root_{banc.version} as the canonical ID column
    id_col = f"root_{args.banc_version}"
    meta = meta.rename(columns={id_col: "root_id"})
    meta = meta[meta["root_id"].notna() & (meta["root_id"] != "")].copy()

    # --- 2. Quality filter ---
    print("Applying quality filter...")
    quality_mask = (
        ~meta["super_class"].str.contains(
            "glia|trachea|not_a_neuron|merge|debris",
            case=False, na=False
        )
        & ~meta["status"].str.contains(
            "GLIA|TRACHEA|NOT_A_NEURON|DEBRIS|MERGE|DELETE",
            na=False
        )
        & (
            (meta["proofread"].astype(str) == "TRUE")
            | (meta["roughly_proofread"].astype(str) == "TRUE")
        )
    )
    meta = meta[quality_mask].copy()
    print(f"  After quality filter: {meta.shape[0]} neurons")

    # --- 3. Assign clustering_set ---
    # Identify neck_connective neurons via super_class (ascending/descending) —
    # backwards-compatible with the now-deprecated region == "neck_connective" filter.
    clustering_regions = ["central_brain", "ventral_nerve_cord"]
    excluded_sc = ["sensory", "motor", "efferent", "afferent", "visceral"]
    excluded_pattern = "|".join(excluded_sc)

    meta["clustering_set"] = None

    # Visual projection / centrifugal
    visual_mask = meta["super_class"].isin(["visual_centrifugal", "visual_projection"])
    meta.loc[visual_mask, "clustering_set"] = "visual"

    # Ascending/descending neurons → neck_connective bucket
    neck_mask = (
        meta["super_class"].str.contains("ascending|descending", na=False)
        & meta["clustering_set"].isna()
    )
    meta.loc[neck_mask, "clustering_set"] = "neck_connective"

    # Region-based, excluding sensory/motor/etc.
    for region in clustering_regions:
        region_mask = (
            (meta["region"] == region)
            & ~meta["super_class"].str.contains(excluded_pattern, na=False)
            & meta["clustering_set"].isna()
        )
        meta.loc[region_mask, "clustering_set"] = region

    clustering_values = ["central_brain", "neck_connective",
                         "ventral_nerve_cord", "visual"]
    cluster_ids = set(
        meta.loc[meta["clustering_set"].isin(clustering_values), "root_id"].values
    )
    print(f"  Neurons in clustering set: {len(cluster_ids)}")

    # --- 4. Filter edgelist + prune ---
    print("Filtering edgelist...")
    el = edgelist[
        (edgelist["count"] >= args.min_connection_strength)
        & (edgelist["pre"] != edgelist["post"])
        & (edgelist["pre"].isin(cluster_ids))
        & (edgelist["post"].isin(cluster_ids))
    ].copy()
    print(f"  Edges after filter: {el.shape[0]}")

    # Iterative pruning to strongly connected component
    print("Pruning to strongly connected component...")
    prev_n = -1
    while True:
        both = set(el["pre"].values) & set(el["post"].values)
        if len(both) == prev_n:
            break
        prev_n = len(both)
        el = el[el["pre"].isin(both) & el["post"].isin(both)]
        print(f"  Pruning: {prev_n} neurons")

    neuron_ids = sorted(both)
    n = len(neuron_ids)
    print(f"  Final: {n} neurons, {el.shape[0]} edges")

    # --- 5. Build sparse adjacency matrix ---
    print("Building adjacency matrix...")
    id_map = {nid: i for i, nid in enumerate(neuron_ids)}

    # Aggregate weights per pre-post pair
    el_agg = el.groupby(["pre", "post"], as_index=False)["count"].sum()
    el_agg.rename(columns={"count": "weight"}, inplace=True)

    row_idx = el_agg["pre"].map(id_map).values
    col_idx = el_agg["post"].map(id_map).values
    weights = el_agg["weight"].values.astype(np.float32)

    adj = coo_matrix((weights, (row_idx, col_idx)), shape=(n, n))
    adj_csc = csc_matrix(adj)

    # Column-normalize
    col_sums = np.array(adj_csc.sum(axis=0)).flatten()
    col_sums[col_sums == 0] = 1
    adj_norm = adj_csc.multiply(1.0 / col_sums)

    # Symmetrize: 0.5 * (A_norm + A_norm^T)
    adj_sym = 0.5 * (adj_norm + adj_norm.T)
    adj_sym = adj_sym.astype(np.float32)
    print("  Matrix built and symmetrized")

    # --- 6. Spectral clustering ---
    print(f"Computing spectral clustering (k={args.cluster_count})...")
    labels, embedding = spectral_clustering_from_adj(
        adj_sym, args.cluster_count, random_state=args.cluster_seed
    )
    for cl in sorted(set(labels)):
        print(f"  Cluster {cl}: {(labels == cl).sum()} neurons")

    # --- 7. UMAP ---
    print("Computing UMAP embedding...")
    umap_coords = perform_umap(
        embedding,
        n_neighbors=100,
        random_state=args.embedding_seed,
    )

    # --- 8. Build result dataframe ---
    result = pd.DataFrame({
        "root_id": neuron_ids,
        "spectral_cluster": labels,
        "umap_x": umap_coords[:, 0],
        "umap_y": umap_coords[:, 1],
    })

    # Join supervoxel_id and position from meta
    meta_cols = meta[["root_id", "supervoxel_id", "position"]].drop_duplicates(
        subset="root_id"
    )
    result = result.merge(meta_cols, on="root_id", how="left")

    # --- 9. Assign cns_network labels ---
    print("Assigning cns_network labels...")
    if args.gcs:
        # Fetch latest meta for cns_network labels.
        # On O2 the versioned save.path already has it; otherwise download
        # from GCS via gsutil.
        o2_meta = os.path.join(
            f"/n/data1/hms/neurobio/wilson/connectomes/banc/banc_{args.banc_version}",
            f"banc_{args.banc_version}_meta.feather",
        )
        if os.path.isfile(o2_meta):
            print(f"  Reading labels from O2: {o2_meta}")
            gcs_meta = pd.read_feather(o2_meta)
        else:
            gcs_path = (
                f"gs://lee-lab_brain-and-nerve-cord-fly-connectome/compiled_data/"
                f"banc_{args.banc_version}/banc_{args.banc_version}_meta.feather"
            )
            local_tmp = os.path.join(
                tempfile.gettempdir(),
                f"banc_{args.banc_version}_meta_gcs.feather",
            )
            print(f"  Downloading {gcs_path} ...")
            subprocess.run(["gsutil", "cp", gcs_path, local_tmp], check=True)
            gcs_meta = pd.read_feather(local_tmp)
        id_col = f"root_{args.banc_version}"
        if id_col in gcs_meta.columns:
            gcs_meta = gcs_meta.rename(columns={id_col: "root_id"})
        st_sub = gcs_meta[["root_id", "cns_network"]].dropna(subset=["cns_network"])
        st_sub = st_sub[st_sub["cns_network"] != ""]
        st_map = dict(zip(st_sub["root_id"], st_sub["cns_network"]))
        print(f"  GCS labels: {len(st_map)} neurons with cns_network")
    else:
        # Use cns_network from the locally loaded meta feather
        st_sub = meta[["root_id", "cns_network"]].dropna(subset=["cns_network"])
        st_sub = st_sub[st_sub["cns_network"] != ""]
        st_map = dict(zip(st_sub["root_id"], st_sub["cns_network"]))

    result["existing_cns_network"] = result["root_id"].map(st_map)

    # Assign cns_network per cluster by Hungarian (bijective) matching.
    # Pair each new spectral_cluster with at most one OLD cns_network label so
    # total neuron overlap on the assignment is maximised. Independent per-
    # cluster majority votes let two new clusters both claim e.g. "right
    # olfactory" while no cluster ends up labelled "left olfactory" —
    # Hungarian forbids that. Clusters left unmatched (all labels claimed
    # elsewhere, or zero overlap with every label) keep their numeric string.
    result["cns_network"] = result["spectral_cluster"].astype(str)
    cluster_ids = sorted(result["spectral_cluster"].unique())
    valid_existing = result["existing_cns_network"].dropna()
    valid_existing = valid_existing[valid_existing != ""]
    old_labels = sorted(valid_existing.unique())

    if len(old_labels) == 0:
        print("  No old cns_network labels available — using cluster numbers")
    else:
        K = len(cluster_ids)
        L = len(old_labels)
        label_to_col = {lbl: j for j, lbl in enumerate(old_labels)}
        # K x L overlap matrix (rows = new clusters, cols = old labels)
        M = np.zeros((K, L), dtype=np.int64)
        for i, cl in enumerate(cluster_ids):
            mask = result["spectral_cluster"] == cl
            ex = result.loc[mask, "existing_cns_network"].dropna()
            ex = ex[ex != ""]
            if len(ex) > 0:
                vc = ex.value_counts()
                for lbl, n in vc.items():
                    if lbl in label_to_col:
                        M[i, label_to_col[lbl]] = n
        # Pad to square so any unmatchable row can take a zero-cost padding col
        N = max(K, L)
        M_sq = np.zeros((N, N), dtype=np.int64)
        M_sq[:K, :L] = M
        row_ind, col_ind = linear_sum_assignment(M_sq, maximize=True)
        # row_ind is 0..N-1 in order; col_ind[i] is the column assigned to row i
        for i, cl in enumerate(cluster_ids):
            j = col_ind[i]
            mask = result["spectral_cluster"] == cl
            if j < L and M[i, j] > 0:
                lbl = old_labels[j]
                result.loc[mask, "cns_network"] = lbl
                print(f"  Cluster {cl} -> '{lbl}' "
                      f"({M[i, j]}/{mask.sum()} neurons)")
            else:
                print(f"  Cluster {cl} -> '{cl}' "
                      f"(no available label — using cluster number)")

    result.drop(columns=["existing_cns_network"], inplace=True)

    # --- 10. Save output ---
    os.makedirs(args.output_dir, exist_ok=True)
    output_file = os.path.join(
        args.output_dir,
        f"spectral_clustering_min_connection_strength_"
        f"{args.min_connection_strength}_banc_version_{args.banc_version}"
        f"_cluster_count_{args.cluster_count}_cluster_seed_"
        f"{args.cluster_seed}_embedding_seed_{args.embedding_seed}"
        f"_{args.source}.csv"
    )

    # Final column order
    result = result[["root_id", "supervoxel_id", "position",
                     "spectral_cluster", "umap_x", "umap_y", "cns_network"]]
    result.to_csv(output_file, index=False)

    elapsed = time.time() - t_start
    print(f"\nSaved: {output_file}")
    print(f"  {result.shape[0]} neurons, "
          f"{result['spectral_cluster'].nunique()} clusters")
    print(f"  Total time: {elapsed / 60:.1f} minutes")

    # --- 11. Interactive UMAP plot ---
    print("Generating interactive UMAP plot...")
    plot_df = result.copy()
    plot_df["spectral_cluster"] = plot_df["spectral_cluster"].astype(str)

    fig = px.scatter(
        plot_df,
        x="umap_x", y="umap_y",
        color="cns_network",
        hover_data=["root_id", "spectral_cluster", "cns_network"],
        title=f"BANC v{args.banc_version} Spectral Clustering UMAP "
              f"(k={args.cluster_count})",
        labels={"umap_x": "UMAP 1", "umap_y": "UMAP 2",
                "cns_network": "CNS Network"},
        width=1000, height=800,
    )
    fig.update_traces(marker=dict(size=2, opacity=0.6))
    fig.update_layout(legend=dict(itemsizing="constant"))

    html_file = os.path.join(
        args.output_dir,
        f"spectral_clustering_umap_banc_version_{args.banc_version}_{args.source}.html"
    )
    fig.write_html(html_file)
    print(f"  UMAP plot saved: {html_file}")


if __name__ == "__main__":
    main()

File: banc/clustering/banc-spectral-clustering.R
#' banc-spectral-clustering — R port of the BANC CNS-network spectral clustering.
#'
#' Column-normalised symmetrised adjacency → normalised Laplacian → KMeans
#' on bottom-k eigenvectors → UMAP. Cluster labels assigned by majority vote
#' against existing SeaTable `cns_network` annotations.
#'
#' @section Reads:
#'   - `banc_<ver>_meta.feather`, `banc_<ver>_edgelist_simple_<src>.feather`
#'   - SeaTable `banc_meta`
#'
#' @section Writes:
#'   - `data/cns_network/spectral_clustering_..._<src>.csv`
#'
#' @section CLI:
#'   --source {v2,v3} --min-connection-strength N --cluster-count K --cluster-seed N --embedding-seed N
#'
#' @section Invoked by:
#'   production v888 rebuild chain.
#'
#' @section Used by:
#'   BANC-project/R/figures/panels_cns_networks.R, panels_cns_network_diagram.R,
#'   panels_cns_network_analyses.R (read `.banc_spectral_csv` for Fig. 6);
#'   R/annotations/banc-spectral-cluster-update.R (pushes cluster labels to SeaTable).
#'
#' @section Schema:
#'   banc_888_cns_network_spectral_clustering_v2.md.
#'
#' @section Paper:
#'   Methods §"Spectral clustering".

###########################################################
### Spectral Clustering of BANC CNS Network
###
### R port of banc-spectral-clustering.py
###
### Performs spectral clustering on the neuron-level
### connectivity graph for central brain, VNC, neck
### connective, and visual neurons.
###
### Algorithm:
###   1. Load BANC meta + simple edgelist from feather (source=v2 or v3)
###   2. Quality filter (exclude glia, trachea, unproofread)
###   3. Assign clustering_set (central brain, VNC, neck, visual)
###   4. Filter edgelist, prune to strongly connected component
###   5. Build column-normalized, symmetrized adjacency matrix
###   6. Spectral clustering (normalized Laplacian, KMeans)
###   7. UMAP of spectral embedding
###   8. Assign cns_network labels from seatable
###   9. Save CSV (output suffix carries the source: _v2 or _v3)
###
### Parameters:
###   min_connection_strength = 1 (paper Methods §"Spectral clustering"; override via --min-connection-strength)
###   cluster_count = 13 (paper Methods §"Spectral clustering"; override via --cluster-count)
###   cluster_seed = 10
###   embedding_seed = 3
###   UMAP: n_neighbors=100, metric=cosine, min_dist=0
###
### Input:
###   - banc_{version}_meta.feather
###   - banc_{version}_edgelist_simple_{v2|v3}.feather
###
### Output:
###   - data/cns_network/spectral_clustering_..._banc_version_{version}_..._{v2|v3}.csv
###   - data/cns_network/seatable_cns_network_snapshot.csv
###
### Usage: Rscript banc/clustering/banc-spectral-clustering.R [--source v2|v3] [--min-connection-strength N]
###########################################################
source("banc/banc-startup.R")

local({
  # Source selection
  .parse_source <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    i <- which(args == "--source")
    if (length(i) == 1 && length(args) >= i + 1) return(tolower(args[i + 1]))
    env <- Sys.getenv("BANC_SYN_SOURCE", unset = NA_character_)
    if (!is.na(env) && nzchar(env)) return(tolower(env))
    if (exists("banc.synapse.source.default")) return(tolower(banc.synapse.source.default))
    "v3"
  }
  syn_source <- .parse_source()
  if (!syn_source %in% c("v2", "v3"))
    stop(sprintf("--source must be 'v2' or 'v3' (got '%s')", syn_source))

  .parse_int_flag <- function(flag, default) {
    args <- commandArgs(trailingOnly = TRUE)
    i <- which(args == flag)
    if (length(i) == 1 && length(args) >= i + 1) return(as.integer(args[i + 1]))
    default
  }

  message(sprintf("### banc: Spectral clustering v%s | source=%s ###",
                  banc.version, syn_source))
  t_start <- Sys.time()

  # ---------------------------------------------------------------
  # Parameters
  # ---------------------------------------------------------------
  min_connection_strength <- .parse_int_flag("--min-connection-strength", 1L)
  cluster_count <- .parse_int_flag("--cluster-count", 13L)
  cluster_seed <- 10
  embedding_seed <- 3
  umap_n_neighbors <- 100
  dataset_version <- as.integer(banc.version)
  id_col <- paste0("root_", dataset_version)
  # Save to bancpipeline AND BANC-project. The latter is the canonical
  # publication location; the former keeps the working copy alongside the
  # script for fast local iteration.
  output_dirs <- c(
    "data/cns_network",
    "/n/data1/hms/neurobio/wilson/banc/BANC-project/data/cns_network"
  )

  # ---------------------------------------------------------------
  # 1. Load data
  # ---------------------------------------------------------------
  message("Loading data...")

  # Auto-detect data directory
  data_candidates <- c(
    file.path("/n/data1/hms/neurobio/wilson/connectomes/banc",
              paste0("banc_", dataset_version)),
    file.path("lee-lab_brain-and-nerve-cord-fly-connectome/compiled_data",
              paste0("banc_", dataset_version))
  )
  data_dir <- NULL
  for (d in data_candidates) {
    if (dir.exists(d)) { data_dir <- d; break }
  }
  if (is.null(data_dir)) stop("Could not find BANC ", dataset_version, " data directory")

  meta <- arrow::read_feather(
    file.path(data_dir, paste0("banc_", dataset_version, "_meta.feather"))
  )
  edgelist <- arrow::read_feather(
    file.path(data_dir, paste0("banc_", dataset_version,
                                "_edgelist_simple_", syn_source, ".feather"))
  )

  # Use the current version's root_{version} column as canonical ID, renamed to root_id
  meta$root_id <- meta[[id_col]]
  meta <- meta[!is.na(meta$root_id) & meta$root_id != "", ]
  message(sprintf("  Meta: %d neurons, Edgelist: %d edges",
                  nrow(meta), nrow(edgelist)))

  # ---------------------------------------------------------------
  # 2. Quality filter
  # ---------------------------------------------------------------
  message("Applying quality filter...")
  meta <- meta %>%
    dplyr::filter(
      !grepl("glia|trachea|not_a_neuron|merge|debris",
             super_class, ignore.case = TRUE),
      !grepl("GLIA|TRACHEA|NOT_A_NEURON|DEBRIS|MERGE|DELETE", status),
      proofread == "TRUE" | roughly_proofread == "TRUE"
    )
  message(sprintf("  After quality filter: %d neurons", nrow(meta)))

  # ---------------------------------------------------------------
  # 3. Assign clustering_set
  # ---------------------------------------------------------------
  # Identify neck_connective neurons via super_class (ascending/descending) —
  # backwards-compatible with the now-deprecated region=="neck_connective" filter.
  clustering_regions <- c("central_brain", "ventral_nerve_cord")
  excluded_sc <- c("sensory", "motor", "efferent", "afferent", "visceral")

  meta <- meta %>%
    dplyr::mutate(
      clustering_set = dplyr::case_when(
        super_class %in% c("visual_centrifugal", "visual_projection") ~ "visual",
        grepl("ascending|descending", super_class) ~ "neck_connective",
        region %in% clustering_regions &
          !grepl(paste(excluded_sc, collapse = "|"), super_class) ~ region,
        TRUE ~ NA_character_
      )
    )

  clustering_values <- c("central_brain", "neck_connective",
                         "ventral_nerve_cord", "visual")
  cluster_ids <- meta %>%
    dplyr::filter(clustering_set %in% clustering_values) %>%
    dplyr::pull(root_id) %>%
    unique()
  message(sprintf("  Neurons in clustering set: %d", length(cluster_ids)))

  # ---------------------------------------------------------------
  # 4. Filter edgelist + prune
  # ---------------------------------------------------------------
  message("Filtering edgelist...")
  el <- edgelist %>%
    dplyr::filter(
      count >= min_connection_strength,
      pre != post,
      pre %in% cluster_ids,
      post %in% cluster_ids
    )
  message(sprintf("  Edges after filter: %d", nrow(el)))

  # Iterative pruning to strongly connected component
  message("Pruning to strongly connected component...")
  prev_n <- -1
  while (TRUE) {
    both <- intersect(unique(el$pre), unique(el$post))
    if (length(both) == prev_n) break
    prev_n <- length(both)
    el <- el %>% dplyr::filter(pre %in% both, post %in% both)
    message(sprintf("  Pruning: %d neurons", prev_n))
  }

  neuron_ids <- sort(both)
  n <- length(neuron_ids)
  message(sprintf("  Final: %d neurons, %d edges", n, nrow(el)))

  # ---------------------------------------------------------------
  # 5. Build sparse adjacency matrix
  # ---------------------------------------------------------------
  message("Building adjacency matrix...")
  id_map <- setNames(seq_along(neuron_ids), neuron_ids)

  # Aggregate weights per pre-post pair
  el_agg <- el %>%
    dplyr::group_by(pre, post) %>%
    dplyr::summarise(weight = sum(count), .groups = "drop")

  i_idx <- id_map[el_agg$pre]
  j_idx <- id_map[el_agg$post]

  adj <- Matrix::sparseMatrix(
    i = i_idx, j = j_idx, x = el_agg$weight,
    dims = c(n, n),
    dimnames = list(neuron_ids, neuron_ids)
  )

  # Column-normalize
  col_sums <- Matrix::colSums(adj)
  col_sums[col_sums == 0] <- 1
  adj_norm <- adj %*% Matrix::Diagonal(n, 1 / col_sums)

  # Symmetrize
  adj_sym <- 0.5 * (adj_norm + Matrix::t(adj_norm))
  message("  Matrix built and symmetrized")

  # ---------------------------------------------------------------
  # 6. Spectral clustering
  # ---------------------------------------------------------------
  message(sprintf("Computing spectral clustering (k=%d)...", cluster_count))

  # Normalized Laplacian: L = I - D^{-1/2} A D^{-1/2}
  deg <- Matrix::rowSums(adj_sym)
  deg[deg == 0] <- 1
  deg_inv_sqrt <- 1 / sqrt(deg)
  D_inv_sqrt <- Matrix::Diagonal(n, deg_inv_sqrt)
  lap <- Matrix::Diagonal(n) - D_inv_sqrt %*% adj_sym %*% D_inv_sqrt

  # Bottom-k eigenvectors
  eig <- RSpectra::eigs_sym(lap, k = cluster_count, which = "SM")
  eigvec <- eig$vectors

  # Row-normalize
  row_norms <- sqrt(rowSums(eigvec^2))
  row_norms[row_norms == 0] <- 1
  embedding <- eigvec / row_norms

  # KMeans
  set.seed(cluster_seed)
  km <- kmeans(embedding, centers = cluster_count, nstart = 25, iter.max = 100)
  labels <- km$cluster  # 1-indexed

  for (cl in sort(unique(labels))) {
    message(sprintf("  Cluster %d: %d neurons", cl, sum(labels == cl)))
  }

  # ---------------------------------------------------------------
  # 7. UMAP embedding
  # ---------------------------------------------------------------
  message("Computing UMAP embedding...")
  umap_result <- uwot::umap(
    embedding,
    n_neighbors = umap_n_neighbors,
    metric = "cosine",
    min_dist = 0,
    n_components = 2,
    seed = embedding_seed,
    n_threads = 1
  )

  # ---------------------------------------------------------------
  # 8. Build result + assign cns_network labels
  # ---------------------------------------------------------------
  result <- data.frame(
    root_id = neuron_ids,
    spectral_cluster = labels,
    umap_x = umap_result[, 1],
    umap_y = umap_result[, 2],
    stringsAsFactors = FALSE
  )

  # Join supervoxel_id and position from meta
  meta_cols <- meta %>%
    dplyr::select(root_id, supervoxel_id, position) %>%
    dplyr::distinct(root_id, .keep_all = TRUE)
  result <- result %>%
    dplyr::left_join(meta_cols, by = "root_id")

  # Get cns_network from seatable and save snapshot for Python
  message("Querying seatable for cns_network labels...")
  st <- tryCatch({
    banctable_query(paste0("SELECT ", id_col, ", cns_network FROM banc"))
  }, error = function(e) {
    message("  Seatable query failed, using cns_network from meta feather")
    meta %>% dplyr::select(root_id, cns_network)
  })

  # Normalise column names
  if (id_col %in% names(st)) {
    names(st)[names(st) == id_col] <- "root_id"
  }
  st <- st %>%
    dplyr::filter(!is.na(cns_network), cns_network != "") %>%
    dplyr::distinct(root_id, .keep_all = TRUE)

  # Save seatable snapshot to bancpipeline (working copy used by downstream
  # scripts and the heatmap). The snapshot is the OLD cns_network values used
  # for the cluster->label majority vote in step 9, so it pairs with the
  # spectral CSV emitted below.
  primary_dir <- output_dirs[1]
  if (!dir.exists(primary_dir)) dir.create(primary_dir, recursive = TRUE)
  snapshot_path <- file.path(primary_dir, "seatable_cns_network_snapshot.csv")
  readr::write_csv(
    st %>% dplyr::rename(!!id_col := root_id),
    snapshot_path
  )
  message(sprintf("  Saved seatable snapshot: %s (%d entries)", snapshot_path, nrow(st)))

  # Assign cns_network per cluster by plain per-cluster majority vote on the
  # OLD SeaTable cns_network labels (neurons missing a label are discounted).
  # Same OLD label may be assigned to multiple clusters when the network
  # genuinely splits a region (e.g., leg VNC into left/right). Clusters with
  # zero overlap with any OLD label keep their numeric cluster string.
  result <- result %>%
    dplyr::left_join(
      st %>% dplyr::select(root_id, existing_cns = cns_network),
      by = "root_id"
    )
  result$cns_network <- as.character(result$spectral_cluster)

  cluster_ids <- sort(unique(result$spectral_cluster))
  old_labels <- sort(unique(result$existing_cns[
    !is.na(result$existing_cns) & result$existing_cns != ""
  ]))

  if (length(old_labels) == 0L) {
    message("  No old cns_network labels available — using cluster numbers")
  } else {
    for (cl in cluster_ids) {
      mask <- result$spectral_cluster == cl
      ex <- result$existing_cns[mask]
      ex <- ex[!is.na(ex) & ex != ""]
      if (length(ex) == 0L) {
        message(sprintf("  Cluster %d -> '%s' (no labelled neurons — using cluster number)",
                        cl, as.character(cl)))
        next
      }
      tab <- sort(table(ex), decreasing = TRUE)
      lbl <- names(tab)[1]
      result$cns_network[mask] <- lbl
      message(sprintf("  Cluster %d -> '%s' (%d/%d neurons, %.1f%% of labelled)",
                      cl, lbl, tab[1], sum(mask),
                      100 * tab[1] / length(ex)))
    }
  }

  result <- result %>% dplyr::select(-existing_cns)

  # ---------------------------------------------------------------
  # 9. Save output
  # ---------------------------------------------------------------
  output_basename <- sprintf(
    "spectral_clustering_min_connection_strength_%d_banc_version_%d_cluster_count_%d_cluster_seed_%d_embedding_seed_%d_%s.csv",
    min_connection_strength, dataset_version, cluster_count,
    cluster_seed, embedding_seed, syn_source)
  # Dated copy preserves historical runs alongside the canonical (overwritten)
  # filename. Date is the calendar date the script started (UTC-naïve, fine for
  # day-granularity ordering).
  dated_basename <- sub("\\.csv$",
                        sprintf("_%s.csv", format(Sys.Date(), "%Y-%m-%d")),
                        output_basename)

  output_df <- result %>%
    dplyr::select(root_id, supervoxel_id, position,
                  spectral_cluster, umap_x, umap_y, cns_network)

  for (od in output_dirs) {
    if (!dir.exists(od)) dir.create(od, recursive = TRUE)
    for (.bn in c(output_basename, dated_basename)) {
      out_path <- file.path(od, .bn)
      readr::write_csv(output_df, out_path)
      message(sprintf("\nSaved: %s (%d neurons, %d clusters)",
                      out_path, nrow(output_df),
                      length(unique(output_df$spectral_cluster))))
    }
  }

  message(sprintf("### banc: Spectral clustering complete [%s] ###",
                  format(round(difftime(Sys.time(), t_start, units = "mins"), 1))))
})
